### PR TITLE
ddc processing + events

### PR DESCRIPTION
**Pull request by aie0**
## Types of Changes
Please select the branch type you are merging and fill in the relevant template.

- [ ] Hotfix
- [ ] Release
- [ ] Fix or Feature

### Types of Changes

- [ ] Tech Debt (Code improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Dependency upgrade (A change in substrate or any 3rd party crate version)

### Migrations and Hooks

- [ ] This change requires a runtime migration.
- [ ] Modifies `on_initialize`
- [ ] Modifies `on_finalize`

### Checklist for Fix or Feature

- [ ] Change has been tested locally.
- [ ] Change adds / updates tests if applicable.
- [ ] Changelog doc updated.
- [ ] `spec_version` has been incremented.
- [ ] `network-relayer`'s [events](https://github.com/Cerebellum-Network/network-relayer/blob/dev-cere/shared/substrate/events.go) have been updated according to the blockchain events if applicable.
- [ ] All CI checks have been passed successfully

## Checklist for Hotfix

- [ ] Changelog has been updated.
- [ ] Crate version has been updated.
- [ ] `spec_version` has been incremented.
- [ ] Transaction version has been updated if required.
- [ ] Pull Request to `dev` has been created.
- [ ] Pull Request to `staging` has been created.
- [ ] `network-relayer`'s [events](https://github.com/Cerebellum-Network/network-relayer/blob/dev-cere/shared/substrate/events.go) have been updated according to the blockchain events if applicable.
- [ ] All CI checks have been passed successfully

## Checklist for Release

- [ ] Change has been deployed to Devnet.
- [ ] Change has been tested in Devnet.
- [ ] Change has been deployed to Qanet.
- [ ] Change has been tested in Qanet.
- [ ] Change has been deployed to Testnet.
- [ ] Change has been tested in Testnet.
- [ ] Changelog has been updated.
- [ ] Crate version has been updated.
- [ ] Spec version has been updated.
- [ ] Transaction version has been updated if required.
- [ ] All CI checks have been passed successfully

### CODE REVIEW

File: pallets/ddc-verification/src/lib.rs [hunk 1]
```diff
@@ -304,24 +336,100 @@ pub mod pallet {
...
+					}
+					None
+				},
+			};
 		}
 	}
 
 	impl<T: Config> Pallet<T> {
+		pub(crate) fn create_merkle_root(leaves: Vec<ActivityHash>) -> ActivityHash {
+			let hash = sp_runtime::traits::BlakeTwo256::trie_root(
```
Comment (aie0) [inline pallets/ddc-verification/src/lib.rs:422]: we need this to be ActivityHasher (let hash = sp_runtime::traits::BlakeTwo256::trie_root() => (let hash = T::ActivityHasher::trie_root()

maybe changing from (type ActivityHasher: frame_support::StorageHasher<Output = ActivityHash>; )to ActivityHasher: sp_runtime::traits::Hasher<Out = ActivityHash>;

to solve this, but everyone should use the same hasher

File: pallets/ddc-verification/src/lib.rs [hunk 2]
```diff
@@ -304,24 +336,100 @@ pub mod pallet {
...
 				Percent::from_percent(T::MAJORITY),
-			);
+			) {
+				Ok(nodes_activity_in_consensus) => {
+					// Process node_activities
+					let mut sorted_activities = nodes_activity_in_consensus.clone();
+					sorted_activities.sort();
+
+					let leaves: Vec<ActivityHash> =
+						sorted_activities.iter().map(|activity| activity.hash::<T>()).collect();
```
Comment (aie0) [inline pallets/ddc-verification/src/lib.rs:392]: maybe worthwhile to use trie. ordered_trie_root and remove this sorting

Comment (ayushmishra2005) [reply pallets/ddc-verification/src/lib.rs:390]: Done

File: pallets/ddc-verification/src/lib.rs (hunk 1 above)
Comment (ayushmishra2005) [reply pallets/ddc-verification/src/lib.rs]: done

File: pallets/ddc-verification/src/lib.rs [hunk 3]
```diff
@@ -68,7 +68,7 @@ pub mod pallet {
 		type WeightInfo: WeightInfo;
 		type ClusterManager: ClusterManager<Self>;
 		type NodeVisitor: NodeVisitor<Self>;
-		type ActivityHasher: StorageHasher<Output = ActivityHash>;
+		type ActivityHasher: Hash<Output = H256>;
```
Comment (aie0) [inline pallets/ddc-verification/src/lib.rs:71]: change to Hash<Output = ActivityHash >;

**Pull request merged**